File: backend/go-chi-hexagonal/internal/application/service/user_service.go
package service

import (
	"context"
	"fmt"

	"github.com/sakrafux/realworld-tech-comparison/backend/go-chi-hexagonal/internal/application/port"
	"github.com/sakrafux/realworld-tech-comparison/backend/go-chi-hexagonal/internal/domain"
)

type userService struct {
	userRepo       port.UserRepository
	passwordHasher port.PasswordHasher
}

func NewUserService(userRepo port.UserRepository, passwordHasher port.PasswordHasher) port.UserService {
	return &userService{
		userRepo:       userRepo,
		passwordHasher: passwordHasher,
	}
}

func (s *userService) Register(ctx context.Context, cmd port.RegisterCommand) (*domain.User, error) {
	// Check if user already exists by email
	existingUser, err := s.userRepo.FindByEmail(ctx, cmd.Email)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if existingUser != nil {
		return nil, domain.NewAlreadyExistsError("Email already exists")
	}

	// Check if user already exists by username
	existingUser, err = s.userRepo.FindByUsername(ctx, cmd.Username)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if existingUser != nil {
		return nil, domain.NewAlreadyExistsError("Username already exists")
	}

	// Hash password
	hashedPassword, err := s.passwordHasher.Hash(cmd.Password)
	if err != nil {
		return nil, domain.NewInternalError(fmt.Sprintf("failed to hash password: %v", err))
	}

	user := &domain.User{
		Username: cmd.Username,
		Email:    cmd.Email,
		Password: hashedPassword,
		Bio:      "",
		Image:    nil,
	}

	if err := s.userRepo.Create(ctx, user); err != nil {
		return nil, domain.NewInternalError(err.Error())
	}

	return user, nil
}

func (s *userService) Login(ctx context.Context, cmd port.LoginCommand) (*domain.User, error) {
	user, err := s.userRepo.FindByEmail(ctx, cmd.Email)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if user == nil {
		return nil, domain.NewNotFoundError("User not found")
	}

	if err := s.passwordHasher.Compare(user.Password, cmd.Password); err != nil {
		return nil, domain.NewInvalidCredentialsError("Invalid email or password")
	}

	return user, nil
}

func (s *userService) GetUser(ctx context.Context, query port.GetUserQuery) (*domain.User, error) {
	user, err := s.userRepo.FindByID(ctx, query.ID)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if user == nil {
		return nil, domain.NewResourceNotFound("User", "id", query.ID)
	}
	return user, nil
}

func (s *userService) UpdateUser(ctx context.Context, cmd port.UpdateUserCommand) (*domain.User, error) {
	user, err := s.userRepo.FindByID(ctx, cmd.ID)
	if err != nil {
		return nil, domain.NewInternalError(err.Error())
	}
	if user == nil {
		return nil, domain.NewResourceNotFound("User", "id", cmd.ID)
	}

	// Check for duplicate email if it's being updated
	if cmd.Email != nil && *cmd.Email != user.Email {
		existingUser, err := s.userRepo.FindByEmail(ctx, *cmd.Email)
		if err != nil {
			return nil, domain.NewInternalError(err.Error())
		}
		if existingUser != nil {
			return nil, domain.NewAlreadyExistsError("Email already exists")
		}
	}

	// Check for duplicate username if it's being updated
	if cmd.Username != nil && *cmd.Username != user.Username {
		existingUser, err := s.userRepo.FindByUsername(ctx, *cmd.Username)
		if err != nil {
			return nil, domain.NewInternalError(err.Error())
		}
		if existingUser != nil {
			return nil, domain.NewAlreadyExistsError("Username already exists")
		}
	}

	var hashedPassword *string
	if cmd.Password != nil {
		hp, err := s.passwordHasher.Hash(*cmd.Password)
		if err != nil {
			return nil, domain.NewInternalError(fmt.Sprintf("failed to hash password: %v", err))
		}
		hashedPassword = &hp
	}

	user.Update(cmd.Username, cmd.Email, cmd.Bio, cmd.Image, hashedPassword)

	if err := s.userRepo.Update(ctx, user); err != nil {
		return nil, domain.NewInternalError(err.Error())
	}

	return user, nil
}

File: backend/typescript-node-express-hive/src/cells/user/user.service.ts
import { User, type Profile } from "./user.domain.js";
import type {
    RegisterCommand,
    LoginCommand,
    UpdateUserCommand,
    UserService,
    UserRepository,
} from "./user.ports.js";
import {
    newAlreadyExistsError,
    newInvalidCredentialsError,
    newNotFoundError,
    newResourceNotFound,
} from "../../shared/errors/app-error.js";
import type { PasswordHasher } from "../../shared/security/password.js";

export class DefaultUserService implements UserService {
    constructor(
        private repo: UserRepository,
        private hasher: PasswordHasher,
    ) {}

    async register(cmd: RegisterCommand): Promise<User> {
        const existingEmail = await this.repo.findByEmail(cmd.email);
        if (existingEmail) {
            throw newAlreadyExistsError("Email already exists");
        }

        const existingUsername = await this.repo.findByUsername(cmd.username);
        if (existingUsername) {
            throw newAlreadyExistsError("Username already exists");
        }

        const hashedPassword = await this.hasher.hash(cmd.password);
        const user = new User(0, cmd.username, cmd.email, hashedPassword, "", null);

        await this.repo.create(user);
        return user;
    }

    async login(cmd: LoginCommand): Promise<User> {
        const user = await this.repo.findByEmail(cmd.email);
        if (!user) {
            throw newNotFoundError("User not found");
        }

        await this.hasher.compare(user.password, cmd.password);
        return user;
    }

    async getUser(id: number): Promise<User> {
        const user = await this.repo.findById(id);
        if (!user) {
            throw newResourceNotFound("User", "id", id);
        }
        return user;
    }

    async getUserByUsername(username: string): Promise<User> {
        const user = await this.repo.findByUsername(username);
        if (!user) {
            throw newResourceNotFound("User", "username", username);
        }
        return user;
    }

    async updateUser(cmd: UpdateUserCommand): Promise<User> {
        const user = await this.getUser(cmd.id);

        if (cmd.email && cmd.email !== user.email) {
            const existingEmail = await this.repo.findByEmail(cmd.email);
            if (existingEmail) {
                throw newAlreadyExistsError("Email already exists");
            }
        }

        if (cmd.username && cmd.username !== user.username) {
            const existingUsername = await this.repo.findByUsername(cmd.username);
            if (existingUsername) {
                throw newAlreadyExistsError("Username already exists");
            }
        }

        let hashedPassword;
        if (cmd.password) {
            hashedPassword = await this.hasher.hash(cmd.password);
        }

        user.update({
            username: cmd.username,
            email: cmd.email,
            bio: cmd.bio,
            image: cmd.image,
            password: hashedPassword,
        });

        await this.repo.update(user);
        return user;
    }

    async getProfile(username: string, observerId?: number): Promise<Profile> {
        const profile = await this.repo.getProfileByUsername(username, observerId);
        if (!profile) {
            throw newResourceNotFound("Profile", "username", username);
        }
        return profile;
    }

    async followUser(followerId: number, username: string): Promise<Profile> {
        const userToFollow = await this.repo.findByUsername(username);
        if (!userToFollow) {
            throw newResourceNotFound("User", "username", username);
        }

        await this.repo.follow(followerId, userToFollow.id);
        return this.getProfile(username, followerId);
    }

    async unfollowUser(followerId: number, username: string): Promise<Profile> {
        const userToUnfollow = await this.repo.findByUsername(username);
        if (!userToUnfollow) {
            throw newResourceNotFound("User", "username", username);
        }

        await this.repo.unfollow(followerId, userToUnfollow.id);
        return this.getProfile(username, followerId);
    }
}
